packages/babel-parser/src/parser/lval.js

```diff
@@ -97,6 +112,7 @@ export default class LValParser extends NodeUtils {
         break;
 
       case "ObjectProperty":
+        // ObjectProperty is not allowed in LHS
```

Comment:
Can't it be something like `({ a: (foo) } = {})`?

Reply:
Good catch. I think we may have to remove error detections on `toAssignable` and make it plainly AST transforms

```js
({ a: (foo) } = {}) => {}
```

If we are good with `(foo)` after `toAssignable(ast`{ a: (foo) } = {}`)`, the error `(foo)` can not populate when we are reinterpreting the assignment expression to assignment pattern, because the `left` of assignment is already an `ArrayPattern`.

We may have to record these errors in a scope chain like we did in #12230 and finally throws when we are sure it is a binding pattern (instead of an LHS).

Reply:
Addressed in 00a776c.